### PR TITLE
fix: check variable essentialness in regions checks

### PR DESCRIPTION
**Pull request by mlutze**
The problem is that the region checks assume that a type is fully minimized, but our minimization technique is not a true minimization, in that irrelevant variables may remain inside. This means that a region variable that is actually irrelevant to the formula may still appear in the formula and therefore be identified as escaping.

I've made a change where we no longer rely on minimization, and we instead explicitly check whether the variable is relevant to the formula. To see if a variable `r` is relevant in a type `t`, I calculate `t0 = t[r -> false]` and `t1 = t[r -> true]`. Then I check if `t0 == t1` modulo Boolean equivalence (where all type variables are rigid).

This seems to work, but we move from a state of "variable elimination order is related to errors" to "variable elimination order is related to Boolean formula explosion"

### CODE REVIEW

**Comment by magnus-madsen** [comment]
> This seems to work, but we move from a state of "variable elimination order is related to errors" to "variable elimination order is related to Boolean formula explosion"

What does this mean? That the check is too expensive?

Can you try to avoid BU and instead check for equivalence of truth tables?

**Comment by mlutze** [comment]
> What does this mean? That the check is too expensive?

Two things:
1. Reversing the order of the variables in BU still has a large effect on the size of formulas, so we have to keep this semi-arbitrary ordering in order to compile.
2. The modification indeed seems expensive in the worst cases.

> Can you try to avoid BU and instead check for equivalence of truth tables?

I'll see if there's a performance impact.

**Comment by magnus-madsen** [comment]
should we enable this together with the other flag?

**Comment by magnus-madsen** [comment]
or perhaps we can make it work and then use a random variable ordering?

**Comment by mlutze** [comment]
> should we enable this together with the other flag?

I'm not convinced that flag helps a lot. It seems to me that rigid variables actually simplify the unification problem, since there's less generality in the formula to be accounted for. Not sure though.

> or perhaps we can make it work and then use a random variable ordering?

1. Random means flaky, which I don't like
2. The variable ordering we use actually has some sense, albeit not by design. Variables are sorted by their IDs, then reversed. This means that the highest IDs are eliminated first. The highest IDs belong to the newest variables, which are those most recently introduced in the `Typer`.

**Comment by magnus-madsen** [comment]
Let me rephrase.

- Does the PR work? I.e. all tests run? (with which variable order?)
- We could add a flag `--Xrandom-var-order` to shuffle to variable order to be used for testing.

**Comment by magnus-madsen** [comment]
By work I mean with little performance impact.

**Comment by mlutze** [comment]
>     * Does the PR work? I.e. all tests run? (with which variable order?)

Yes, with `reverse` order. (I.e., I did not change the order in this PR.

>     * We could add a flag `--Xrandom-var-order` to shuffle to variable order to be used for testing.

OK

> By work I mean with little performance impact 

The performance impact is now small enough that I cannot measure it myself, so we'll see with the build server.

> (and with --Xflex-regions or whatever).

Boolean explosion on on `Test.Exp.Regions`. Everything else is fine (except RegionVarEscapes tests)

**Comment by magnus-madsen** [comment]
> Boolean explosion on on Test.Exp.Regions. 

So these time out?

> Everything else is fine (except RegionVarEscapes tests)

And these fail?

**Comment by mlutze** [comment]
> > Boolean explosion on on Test.Exp.Regions.
> 
> So these time out?

I run out of heap space.
 
> > Everything else is fine (except RegionVarEscapes tests)
> 
> And these fail?

Yes

**Comment by magnus-madsen** [comment]
If we ignore those test suites are we then able to randomize the order or no?

**Comment by JonathanStarup** [comment]
> To see if a variable r is relevant in a type t, I calculate t0 = t[r -> false] and t1 = t[r -> true]. Then I check if t0 == t1 modulo Boolean equivalence 

This is quadratic in variables right? I think to improve performance in the happy path, this should only happen if the occurrence check fails.

**Comment by magnus-madsen** [comment]
> > To see if a variable r is relevant in a type t, I calculate t0 = t[r -> false] and t1 = t[r -> true]. Then I check if t0 == t1 modulo Boolean equivalence
> 
> This is quadratic in variables right? I think to improve performance in the happy path, this should only happen if the occurrence check fails.

Its exponential, hence it would make sense to do quick check first and then only do the expensive check.

**Comment by mlutze** [comment]
> > > To see if a variable r is relevant in a type t, I calculate t0 = t[r -> false] and t1 = t[r -> true]. Then I check if t0 == t1 modulo Boolean equivalence
> > 
> > 
> > This is quadratic in variables right? I think to improve performance in the happy path, this should only happen if the occurrence check fails.
> 
> Its exponential, hence it would make sense to do quick check first and then only do the expensive check.

Yes, this check is included in the implementation.

> If we ignore those test suites are we then able to randomize the order or no?

Testing with non-reversed order also hangs on `TestIterator`, I think.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala [hunk 1]
```diff
@@ -421,16 +422,51 @@ object Regions {
...
+      !sameType(t0, t1)
     }
+  }
 
-    // Otherwise return success.
-    ().toSuccess
+  /**
+    * Returns true iff the two types denote the same Boolean function, using the same variables.
+    */
+  private def sameType(t1: Type, t2: Type)(implicit flix: Flix): Boolean = {
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala:464]: I was hoping to avoid extra complexity but I've hacked together a truth-tabley implementation to see what that does.

[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala:464]: The hacky truth-tabley implementation is no good since we are checking whether `r` is relevant to `t` where `t` is any kind of type, not just `Bool`.

I think we can do this by extracting all the boolean types from the type and seeing if `r` is relevant to them

[the remark this answers is not in the document]
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Regions.scala:478]: Its fun that I'm exactly looking at this case in the region proof

**Comment by magnus-madsen** [comment]
What's the current status on this?

**Comment by mlutze** [comment]
Things seem to work. I'll just add documentation and clean it up.

**Comment by mlutze** [comment]
I didn't add the `--random var order` though. I can do that in a separate PR if it's needed.

File: main/src/ca/uwaterloo/flix/language/phase/Regions.scala [hunk 2]
```diff
@@ -421,16 +422,84 @@ object Regions {
...
-    // Return an error if a region variable escapes.
-    if (escapes.nonEmpty) {
-      val rvar = escapes.head
-      return TypeError.RegionVarEscapes(rvar, tpe, loc).toFailure
+  /**
+    * Returns true iff the type variable `tvar` is essential to the boolean formula `tpe`.
+    * Assumes that `tvar` is present in the type.
+    */
+  def essentialToBool(tvar: Type.Var, tpe: Type)(implicit flix: Flix): Boolean = {
+    val t0 = tpe.map {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Regions.scala:456]: This was not clear to me, maybe we can write a small function substituteVar to make it clear.

**Comment by magnus-madsen** [comment]
Once this is merged, what are the next steps?

**Comment by mlutze** [comment]
We should not have any more false errors of that variety. So what remains is the performance/overflow issue that I don't yet know how to solve.

**Comment by mlutze** [comment]
We should also be able to follow up by introducing region syns instead of rigid vars
